### ISSUE TITLE
[question] Changed revision of a transitive dep breaks the binary package selection [conan 2]

### ISSUE THREAD

**Issue by forry** [conan-io/conan#16948]
### What is your question?

Hi,
in our project with conan 2 we have lots of dependencies. It was all working a couple of weeks ago, until some colleagues that postponed the conan 2 installation until now, started to complain about the conan not delivering the binaries for the packages others have. From my investigation, it seems that we have one of our private packages (cwexchange) and some of the conan center ones (assimp) depending on the bzip2/1.0.8. Which lately probably changed the package revision. Thus all the people that did the conan install before the change are fine but for the new ones this is how the graph explain looks like:
```
======== Closest binaries ========
cwexchange/2.3.1@lexocad-external/prebuild
  cwexchange/2.3.1@lexocad-external/prebuild#883e268a7100057e5fb9558a624fbc7e%1723461019.688 (2024-08-12 11:10:19 UTC)
    cwexchange/2.3.1@lexocad-external/prebuild#883e268a7100057e5fb9558a624fbc7e:8e3e772472b55df513b63da598c589d48ad19ac7
      remote: lxdevelop
      settings: Windows, x86_64, Release, msvc, dynamic, Release, 193
      options: enable_tests=False
      requires: boost/1.78.Z@lexocad/prebuild, bzip2/1.0.8#457c272f7da34cb9c67456dd217d36c4:67bfcb7b4b78262b9d05495e479dcd92f747316b, fbxsdk/2017.0.Z@cwexchange/prebuild, functionalplus/0.2.24#af575993474b27ecc8af13a51a304880:da39a3ee5e6b4b0d3255bfef95601890afd80709, glm/0.9.9.6@lexocad/prebuild#a4ffac5cf104e17074a1eb237cb77db4:da39a3ee5e6b4b0d3255bfef95601890afd80709, minizip/1.2.13#75d20b6a37ef70126334467f61acadde:63ed532a52a9463b677fe79dd234080cbbfc0acb, pvrtextool/5.6.Z@cwexchange/prebuild, zlib/1.2.11#fca992a7d96a1b92bd956caa8a97d18f:7bfde258ff4f62f75668d0896dbddedaa7480a0f
      diff
        dependencies
          expected: bzip2/1.0.8#d00dac990f08d991998d624be81a9526:67bfcb7b4b78262b9d05495e479dcd92f747316b, minizip/1.2.13#75d20b6a37ef70126334467f61acadde:3ac988588a966957f006a80bae5ba9f36cb23bdc
          existing: bzip2/1.0.8#457c272f7da34cb9c67456dd217d36c4:67bfcb7b4b78262b9d05495e479dcd92f747316b, minizip/1.2.13#75d20b6a37ef70126334467f61acadde:63ed532a52a9463b677fe79dd234080cbbfc0acb
        explanation: This binary was built with different settings.
```
On our server in the conaninfo.txt for the cwe there is truly the full bzip2 with the package revision but it is not a direct dependency of ours more like it is a transitive dependency of a minizip which is a dependency of assimp and cwexchange. If you have the cache from time before the new package revision of bzip2 and you delete the cwexchange it will still request and download the right binary but if you wipe the cache and start downloading all the packages from scratch you get the missing binary error. So it's like the assimp then probably want the newer one and the cwe still somehow want the older one - but it's the revision, not the version. The bzip2 is a static library, assimp is also a static library but cwe is a shared library.

I don't know what to do with that, I don't even know if my explanation is clear because I don't understand how this could happen when nothing has changed on our part be that the conanfile.py that requests the packages or our private conan2 repository. This is very frustrating, help is appreciated.

### Have you read the CONTRIBUTING guide?

- [ ] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @forry 

Thanks for your question

> I don't know what to do with that, I don't even know if my explanation is clear because I don't understand how this could happen when nothing has changed on our part be that the conanfile.py that requests the packages or our private conan2 repository. This is very frustrating, help is appreciated.

Have you checked this page in the docs? https://docs.conan.io/2/devops/using_conancenter.html
To use packages from ConanCenter there are 2 different recommended approaches:
- Use lockfiles, to guarantee that the versions and revisions that you use in your builds do not change
- Build your own binaries from ``conan-center-index`` source and upload them to your own server too.

It is expected that if new versions (there are requirements using version-ranges in ConanCenter) or new recipe-revisions happen in ConanCenter, and these new versions or revisions are installed, they might need the rebuild, building from source the consumers of those packages.

This happens specially in "embedding" scenarios, and from the output above:
```
bzip2/1.0.8#457c272f7da34cb9c67456dd217d36c4:67bfcb7b4b78262b9d05495e479dcd92f747316b
```
this means that ``cwexchange`` has an "embedding" relation to the ``bzip2``, which could happen in different scenarios, for example if ``cwexchange`` is a ``package_type="application"``, as changing the dependencies, even if they are transitive, that are static libraries need to re-build ``cwexchange``.

To understand why this would be happening, it would be necessary to know the dependency graph, the relations and package-types of related packages, maybe this section of the docs help: https://docs.conan.io/2/reference/binary_model/dependencies.html

Still, even if understanding the dependency graph, the right approach to handle this is the recommendation pointed out above in https://docs.conan.io/2/devops/using_conancenter.html

Btw, the message ``explanation: This binary was built with different settings.`` looks wrong, are you sure this is the exact message that is being printed?

**Comment by forry** [comment]
Thanks, I'll look into more details about lock files. At first glance, the policy to have as much packages from the public repo as possible looked very juicy.

>Btw, the message explanation: This binary was built with different settings. looks wrong, are you sure this is the exact message that is being printed?

Don't worry, the output is doctored because I used a different compiler version when reproducing the message. The M$ still haven't resolved their bug with choosing a msvc toolset version from cmake. I deleted that settings diff since it was irrelevant to my colleague's case and 194->193 compiler version is the only one correctly handled by the out-of-the-box compatibility plugin.

**Comment by memsharded** [comment]
Sounds good, don't hesitate to create new tickets for any further question you might have about lockfiles (or any other topic).

> Don't worry, the output is doctored because I used a different compiler version when reproducing the message.

Great, I thought there was some bug in the ``conan graph explain``.


Thanks very much for the feedback!
